pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/SinkConfigUtils.java

```diff
@@ -318,79 +322,89 @@ public static ExtractedSinkDetails validate(SinkConfig sinkConfig, Path archiveP
             throw new IllegalArgumentException("Sink timeout must be a positive number");
         }
 
-        String sinkClassName;
-        final Class<?> typeArg;
-        final ClassLoader classLoader;
-        if (!isEmpty(sinkConfig.getClassName())) {
-            sinkClassName = sinkConfig.getClassName();
-            // We really don't know if we should use nar class loader or regular classloader
-            ClassLoader jarClassLoader = null;
-            ClassLoader narClassLoader = null;
-            try {
-                jarClassLoader = FunctionCommon.extractClassLoader(archivePath, uploadedInputStreamAsFile);
-            } catch (Exception e) {
+
+        Class<?> typeArg;
+        ClassLoader classLoader;
+        String sinkClassName = sinkConfig.getClassName();
+        ClassLoader jarClassLoader = null;
+        ClassLoader narClassLoader = null;
+        try {
+            jarClassLoader = FunctionCommon.extractClassLoader(archivePath, sinkPackageFile);
+        } catch (Exception e) {
```

Comment:
I would suggest keeping the exception in a variable and same for NarClassLoader. Because if we encountered both exceptions, we should log these exceptions and include them in the final exception this function throw.

Otherwise people will not know why their sink package doesn't work and pulsar doesn't given any useful information.

Reply:
@sijie 

```
jarClassLoader = FunctionCommon.extractClassLoader(archivePath, sinkPackageFile);
```
and 

```
narClassLoader = FunctionCommon.extractNarClassLoader(archivePath, sinkPackageFile);
```

are checks to determine if the package is a JAR or a NAR.  The errors there are not meaningful to the user. 

We have to use the logic after it to determine if the what the user submitted is valid e.g. If a JAR is submitted, a className must also be provided or if className is not provided then the archive needs to be a NAR

Reply:
I understand the logic there. My point is if a user submitted a corrupted jar or nar package, there is no way to tell if it is because jar or nar has problems. My comment is to log the errors when we can't retrieve either class loaders.

Reply:
@sijie we can log the errors but they might mislead users.   For example if a user submitted a JAR, there will be an error loading it as a NAR but why log there error as its not relevant.  Also,  it is inherently difficult to tell whether the file submitted by the user is a JAR or a NAR, unless we assume based on file suffix but currently we do not do that.  The UrlClassloader in Java doesn't throw any exceptions if the file is not a JAR.  Technically you can put anything on the classpath regardless of whether is a JAR or not.  

I rather not log errors that might not have any thing to do with what the

Reply:
@jerrypeng 

first of all, it is going to become very hard to help people to understand the problem if we don't log any errors. based on the experiences that I have helped with pulsar users, one of the problems is that we don't have enough logging information for people to understand what is wrong. I ended up asking people adding logging themselves, recompiling the codebase and rerun the program again. This is an inefficient approach for helping our pulsar users. from this perspective, I would rather having more information than having limited information, especially submitting a pulsar function is an end-user facing operation. 

secondly, I agreed that it probably doesn't tell you if it is a JAR or NAR. but the user who submitted this package, it knows what type is the package. So when the log tells him both NAR and JAR class loader can't load the package, he knows where to look into the problem. Also there can be errors not just related to class loading but other problems (such as malformed url, file not found, filesystem io problem). The stack trace will tell you these problems.

Reply:
@sijie 

> first of all, it is going to become very hard to help people to understand the problem if we don't log any errors.

We are logging errors in the code. Please read further down the code.

The JAR classloader will also be non-null.  The only errors can happen when trying to load as a NAR.  Let just say we log these NAR errors regardless of whether a user submitted a NAR of not.  Later and cluster admin comes back and looks at the logs and see all these errors loading a NAR. At in the best case scenario the admin will just be confused and at worse help him to misdiagnose the problem and perform some action to make things worse.  

Logging misleading information will just create more confusion.  Perhaps the logic of this code

Reply:
If you are just logging the exceptions, they are probably misleading. but if you write the message in a way that provides guidelines, that would be helpful. so avoid misleading information is not to provide anything information, is to provide information with guidelines.

for example, if I have a correct NAR package, but I provide a wrong path (i.e. a typo in the url), what happened? do I just get an exception with "Invalid source package"? The information is not enough, no? because the exceptions are shallowed in both class loaders.

Reply:
@sijie 

> for example, if I have a correct NAR package, but I provide a wrong path (i.e. a typo in the url), what happened? do

That is checked before this logic here occurs.  Either the user uploaded a package or the user provided a link, both cases are handled prior to the classloading logic.  

> do I just get an exception with "Invalid source package"? The information is not enough, no? because the exceptions are shallowed in both class loaders.

The reason the error message is "Invalid source package" is because if we arrive in that part of the logic, we know that something is not right but it could be couple things:

1. The user submitted a JAR but did not provide a classname
2. The user submitted a NAR that is incorrectly formatted

we can't exactly determine which one of the above 2 is the problem because we are not currently assuming if a package is a JAR or a NAR and there is not way for us to determine that.  Perhaps in the future the better logic to have to for us to assume if a package is a JAR or a NAR based on file suffix.  I think that is a reasonable assumption to make and will also make the logic and error handling simpler and we will be able to return more sensible errors to the user.

Reply:
> That is checked before this logic here occurs. Either the user uploaded a package or the user provided a link, both cases are handled prior to the classloading logic.

My example is probably not a good one. because the path checking was done before class loading. 

My point is there are many possibilities that exceptions can be thrown during classloading. These exceptions can be any IO exceptions or other exception when loading the classes. With your code change, these exceptions are dropped and there is no useful logging recorded for the users to understand what went wrong. 

For example, there are many IO operations in a NAR class loader (e.g. unpacking, listing files and etc). These IO operations can throw exceptions at any time (e.g. disk corrupted, disk full and many other stuffs). Even you submitted a good NAR package, but it can still fail. If we don't log the messages, how can people troubleshoot the problem? 

> we can't exactly determine which one of the above 2

If we can't determine, why not just giving out the two errors and let users to know both doesn't work and what are the errors for each case. 

I don't really understand why that would be confusing. You can write the error message like:

"Pulsar can't not determine the package is a NAR package or JAR package. Pulsar attempts to load it as a NAR package but here is the error ...; Pulsar attempts to load it as a JAR package but there is an error. Please check the package and fix the errors."

If you write an error message like above, it is more informative than giving nothing to the users.

Reply:
> These exceptions can be any IO exceptions or other exception when loading the classes. With your code change, these exceptions are dropped and there is no useful logging recorded for the users to understand what went wrong.

My change doesn't drop any exceptions and error messages except for the one that is not correct

> If we can't determine, why not just giving out the two errors and let users to know both doesn't work and what are the errors for each case.

Sure I can add something like "Pulsar can't not determine the package is a NAR package or JAR package. Pulsar attempts to load it as a NAR package but here is the error ...;

Reply:
> My change doesn't drop any exceptions and error messages except for the one that is not correct

Okay. I rechecked the logic. The original code already swallowed the exceptions. but anyway since you are touching this logic, let's improve it with more informative logging.

Reply:
To be honest, I think we should we just assume a user is trying to submit a JAR if the package submitted has a ".jar" suffix and the same for a NAR.  This allows our error handling to be much more clear.

Reply:
I am not sure we should use suffix for this. These days more and more applications don't use suffix for checking if the file is the expected type.